clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java

```diff
@@ -100,6 +100,13 @@ public NavigableMap<Integer, Object> read(ByteBuffer buffer) {
             }
             prevTag = tag;
             int size = ByteUtils.readUnsignedVarint(buffer);
+            if (size < 0 && isNullable())
+                return null;
+            else if (size < 0)
+                throw new SchemaException("field size " + size + " cannot be negative");
+            if (size > buffer.remaining())
+                throw new SchemaException("Error reading field of size " + size + ", only " + buffer.remaining() + " bytes available");
```

Comment:
This validation needs to be performed for every Type of subclass of `DocumentedType`. We can potentially abstract these checks out to a method. 

I would suggest the following:

1. In the base class  `DocumentedType` don't make read() an abstract method. 
2. The read() method in `DocumentedType` should look like this:
```
read() {
    // this method is implemented in child classes
    calculateSizeToRead()
    // perform size validation
    if (size < 0)
    .. blah blah.. // perform the if else size validation you are doing here
    // call child class method to actually read
    _read() 
}

```

This would ensure that if a new `type` is added in future, the author won't forget to perform size checks.

Reply:
That's a good point. I'll work on it. Thanks.

Reply:
@divijvaidya , I was following your suggestion to do the documentType refactor like this https://github.com/apache/kafka/compare/trunk...showuon:taggedFields_temp?expand=1 , and then, I found in `taggedFields`, the size validation is in nested way. So, that makes the validation framework doesn't be that useful. Besides, after the change, the child of `DocumentType` can still override the `read` method if it wants (inherited from parent `Type` class), which makes it not useful. So I reverted back to the original change, and add a comment to the `read` method. Does that make sense?